sycl/include/sycl/detail/common.hpp

```diff
@@ -119,6 +119,58 @@ struct code_location {
 #define _CODELOCFW(a)
 #endif
 
+/// @brief Data type that manages the code_location information in TLS
+/// @details As new SYCL features are added, they all enable the propagation of
+/// the code location information where the SYCL API was called by the
+/// application layer. In order to facilitate this, the tls_code_loc_t object
+/// assists in managing the data in TLS :
+///   (1) Populate the information when you at the top level function in the
+///   call chain. This is usually the end-user entry point function into SYCL.
+///   (2) Remove the information when the object goes out scope in top level
+///   function.
+///
+/// Usage:-
+///   void bar() {
+///     tls_code_loc_t p;
+///     // Print the source information of where foo() was called in main()
+///     std::cout << p.query().fileName() << ":" << p.query().lineNumber() <<
+///     std::endl;
+///   }
```

Comment:
Assuming `bar` is at the ABI boundary, I am still missing how this is better than

```
// TODO: Remove at the next ABI break window.
void bar() [[deprecated]] {
  bar(code_location{});
}
void bar(const code_location &) {
 // Actual implementation.
}
```

So far, this is my biggest concern about this patch.

Reply:
The initial implementation of tracing did this and we had to update the signature at every layer in the stack. The TLS approach allows us to bypass and also apply to SW layers beyond SYCL call chain.